<a href="https://colab.research.google.com/github/mauloredo/Challenge_Telecom_X2/blob/main/x2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PRODUCCIÓN

## CONTROL DE VERSIONES DEL DATAFRAME

**datos_x2**  Dataframe importado del projecto anterior

**datos** Dataframe con columnas object a category, como estaban originalmente

## CONTROL DE OBJETOS

## DESCRIPCIÓN DEL DESAFÍO



Telecom X – Parte 2: Predicción de Cancelación (Churn)

📣 Historia del Desafío

¡Felicidades! 🎉 Has sido promovido después de tu excelente desempeño en el análisis exploratorio de la cancelación de clientes en Telecom X. Tu dedicación, claridad al comunicar los datos y visión estratégica marcaron la diferencia.

Ahora, ¡has sido invitado oficialmente a formar parte del equipo de Machine Learning de la empresa!

🎯 Misión

Tu nueva misión es desarrollar modelos predictivos capaces de prever qué clientes tienen mayor probabilidad de cancelar sus servicios.

La empresa quiere anticiparse al problema de la cancelación, y te corresponde a ti construir un pipeline robusto para esta etapa inicial de modelado.

🧠 Objetivos del Desafío

    Preparar los datos para el modelado (tratamiento, codificación, normalización).

    Realizar análisis de correlación y selección de variables.

    Entrenar dos o más modelos de clasificación.

    Evaluar el rendimiento de los modelos con métricas.

    Interpretar los resultados, incluyendo la importancia de las variables.

    Crear una conclusión estratégica señalando los principales factores que influyen en la cancelación.

🧰 Lo que vas a practicar

✅ Preprocesamiento de datos para Machine Learning
✅ Construcción y evaluación de modelos predictivos
✅ Interpretación de resultados y entrega de insights
✅ Comunicación técnica con enfoque estratégico

🚀 Ahora eres: Analista Junior de Machine Learning

Telecom X confía en tu entrega para dar los próximos pasos hacia una solución de inteligencia predictiva eficaz. ¡Buena suerte!
Comentarios y Actividad

## 1. CARGA Y EXPLORACIÓN INICIAL DE DATOS



    


    1.3 Realizar un análisis exploratorio básico (distribuciones, outliers, estadísticas descriptivas).

### 1.1 Importar el dataset desde el repositorio.

In [14]:
import requests
from io import BytesIO
import pandas as pd

raw_url = 'https://raw.githubusercontent.com/mauloredo/Challenge_Telecom_X2/refs/heads/main/data_x2.csv'

print(f"Attempting to download data from: {raw_url}")

try:
    response = requests.get(raw_url)
    response.raise_for_status()

    datos_x2 = pd.read_csv(BytesIO(response.content))

    print("File downloaded and loaded successfully!")
    print("DataFrame shape:", datos_x2.shape)
    print("\nDataFrame head:")
    print(datos_x2.head())

except requests.exceptions.RequestException as e:
    print(f"Error downloading the file: {e}")
    print("Please ensure the raw URL is correct and the file exists in the repository.")

Attempting to download data from: https://raw.githubusercontent.com/mauloredo/Challenge_Telecom_X2/refs/heads/main/data_x2.csv
File downloaded and loaded successfully!
DataFrame shape: (7043, 21)

DataFrame head:
   Churn  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  False  Female          False     Yes        Yes       9          Yes   
1  False    Male          False      No         No       9          Yes   
2   True    Male          False      No         No       4          Yes   
3   True    Male           True     Yes         No      13          Yes   
4   True  Female           True     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0            No             DSL             No  ...               No   
1           Yes             DSL             No  ...               No   
2            No     Fiber optic             No  ...              Yes   
3            No     Fiber optic             No  

### 1.2 Verificar dimensiones, tipos de datos y valores nulos.


In [15]:
# Verificar dimensiones del DataFrame
print(f"Filas: {datos_x2.shape[0]} | Columnas: {datos_x2.shape[1]}")

# Mostrar nombres de columnas
print("\n📋 Lista de columnas:")
print(datos_x2.columns.tolist())

# Tipos de datos por columna
print("\n🔧 Tipos de datos:")
print(datos_x2.dtypes)

# Recuento de valores nulos por columna
print("\n🧯 Valores nulos por columna:")
print(datos_x2.isnull().sum())

# Verificar valores únicos por columna (ideal para detectar categóricas)
print("\n🔎 Valores únicos por columna:")
print(datos_x2.nunique())

# Estadísticas descriptivas de columnas numéricas
print("\n📊 Estadísticas numéricas:")
print(datos_x2.describe())

# Estadísticas para columnas categóricas (si existen)
print("\n📊 Estadísticas categóricas:")
print(datos_x2.describe(include='object'))



Filas: 7043 | Columnas: 21

📋 Lista de columnas:
['Churn', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Cuentas_diarias']

🔧 Tipos de datos:
Churn                  bool
gender               object
SeniorCitizen          bool
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Cuentas_diarias     float64
dtype: object



Después del análisis exploratiorio se encuentra que los tipos de variables del dataframe difieren con el archivo original, Puede ser que en la importación

Data columns (total 20 columns):
 ind   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   Churn             7043 non-null   bool    
 1   gender            7043 non-null   category
 2   SeniorCitizen     7043 non-null   bool    
 3   Partner           7043 non-null   category
 4   Dependents        7043 non-null   category
 5   tenure            7043 non-null   int64   
 6   PhoneService      7043 non-null   category
 7   MultipleLines     7043 non-null   category
 8   InternetService   7043 non-null   category
 9   OnlineSecurity    7043 non-null   category
 10  OnlineBackup      7043 non-null   category
 11  DeviceProtection  7043 non-null   category
 12  TechSupport       7043 non-null   category
 13  StreamingTV       7043 non-null   category
 14  StreamingMovies   7043 non-null   category
 15  Contract          7043 non-null   category
 16  PaperlessBilling  7043 non-null   category
 17  PaymentMethod     7043 non-null   category
 18  MonthlyCharges    7043 non-null   float64
 19  TotalCharges      7043 non-null   object  

Churn                  bool
gender               object
SeniorCitizen          bool
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Cuentas_diarias     float64

Por el momento, las columnas se quedarán así ya qeu mas adelante, si decidimos incluirlas para el análisis, se convertirán con alguno de los métodos One-Hot o Label Encoder. Sin embargo se revisan los datos de estas columnas a continuación.

In [16]:
# Obtener columnas categóricas
cat_cols = datos_x2.select_dtypes(include='object').columns

# Obtener número total de filas
total_filas = datos_x2.shape[0]

for col in cat_cols:
    print(f"\n📂 Columna: {col}")

    # Mostrar valores únicos
    uniques = datos_x2[col].unique()
    print("Valores únicos:", uniques)
    print("Cantidad de únicos:", datos_x2[col].nunique())

    # Mostrar frecuencia de cada valor
    value_counts = datos_x2[col].value_counts(dropna=False)
    print("\n📊 Frecuencia de cada valor:")
    print(value_counts)

    # Sumar total de frecuencias
    suma_total = value_counts.sum()
    print(f"\n🧾 Suma total de valores: {suma_total}")

    # Validar coincidencia con número de filas
    if suma_total == total_filas:
        print("✅ La suma de frecuencias coincide con el total de filas.")
    else:
        print("⚠️ La suma NO coincide con el total de filas. Revisar posibles errores o valores perdidos.")





📂 Columna: gender
Valores únicos: ['Female' 'Male']
Cantidad de únicos: 2

📊 Frecuencia de cada valor:
gender
Male      3555
Female    3488
Name: count, dtype: int64

🧾 Suma total de valores: 7043
✅ La suma de frecuencias coincide con el total de filas.

📂 Columna: Partner
Valores únicos: ['Yes' 'No']
Cantidad de únicos: 2

📊 Frecuencia de cada valor:
Partner
No     3641
Yes    3402
Name: count, dtype: int64

🧾 Suma total de valores: 7043
✅ La suma de frecuencias coincide con el total de filas.

📂 Columna: Dependents
Valores únicos: ['Yes' 'No']
Cantidad de únicos: 2

📊 Frecuencia de cada valor:
Dependents
No     4933
Yes    2110
Name: count, dtype: int64

🧾 Suma total de valores: 7043
✅ La suma de frecuencias coincide con el total de filas.

📂 Columna: PhoneService
Valores únicos: ['Yes' 'No']
Cantidad de únicos: 2

📊 Frecuencia de cada valor:
PhoneService
Yes    6361
No      682
Name: count, dtype: int64

🧾 Suma total de valores: 7043
✅ La suma de frecuencias coincide con el total d

In [17]:
print(type(datos_x2))

<class 'pandas.core.frame.DataFrame'>


In [18]:
print("Shape:", datos_x2.shape)
print("¿Está vacío?:", datos_x2.empty)

Shape: (7043, 21)
¿Está vacío?: False


In [20]:
# Transformación de tipos y renombrado del DataFrame

# 1. Convertir columnas object a category en datos_x2
object_columns = datos_x2.select_dtypes(include=['object']).columns
for col in object_columns:
    datos_x2[col] = datos_x2[col].astype('category')

# 2. Renombrar el DataFrame a 'datos'
datos = datos_x2.copy()

# Verificación
print("\nTipos de datos después de la conversión:")
print(datos.dtypes)
print("\nNombre del DataFrame:", [name for name in globals() if globals()[name] is datos])

# 3. Guardar el DataFrame procesado
output_filename = 'datos.csv'
datos.to_csv(output_filename, index=False)
print(f"\nDataFrame guardado como '{output_filename}'")




Tipos de datos después de la conversión:
Churn                   bool
gender              category
SeniorCitizen           bool
Partner             category
Dependents          category
tenure                 int64
PhoneService        category
MultipleLines       category
InternetService     category
OnlineSecurity      category
OnlineBackup        category
DeviceProtection    category
TechSupport         category
StreamingTV         category
StreamingMovies     category
Contract            category
PaperlessBilling    category
PaymentMethod       category
MonthlyCharges       float64
TotalCharges         float64
Cuentas_diarias      float64
dtype: object

Nombre del DataFrame: ['datos']

DataFrame guardado como 'datos.csv'


###  1.3 Realizar un análisis exploratorio básico (distribuciones, outliers, estadísticas descriptivas).

In [23]:
# Estadísticas descriptivas para variables numéricas
print(datos[['tenure', 'MonthlyCharges', 'TotalCharges', 'Cuentas_diarias']].describe())

# Estadísticas para variables categóricas
print("\nDistribución de categorías:")
for col in datos.select_dtypes(include=['category', 'bool']).columns:
    print(f"\n{col}:")
    print(datos[col].value_counts(normalize=True))


            tenure  MonthlyCharges  TotalCharges  Cuentas_diarias
count  7043.000000     7043.000000   7043.000000      7043.000000
mean     32.371149       64.761692   2279.734304         2.158723
std      24.559481       30.090047   2266.794470         1.003002
min       0.000000       18.250000      0.000000         0.608333
25%       9.000000       35.500000    398.550000         1.183333
50%      29.000000       70.350000   1394.550000         2.345000
75%      55.000000       89.850000   3786.600000         2.995000
max      72.000000      118.750000   8684.800000         3.958333

Distribución de categorías:

Churn:
Churn
False    0.73463
True     0.26537
Name: proportion, dtype: float64

gender:
gender
Male      0.504756
Female    0.495244
Name: proportion, dtype: float64

SeniorCitizen:
SeniorCitizen
False    0.837853
True     0.162147
Name: proportion, dtype: float64

Partner:
Partner
No     0.516967
Yes    0.483033
Name: proportion, dtype: float64

Dependents:
Dependents
No 

# PRESENTACIÓN

## ESTRUCTURA DEL PROYECTO

1. Carga y Exploración Inicial de Datos

    1.1 Importar el dataset desde el repositorio.

    1.2 Verificar dimensiones, tipos de datos y valores nulos.

    1.3 Realizar un análisis exploratorio básico (distribuciones, outliers, estadísticas descriptivas).

2. Preprocesamiento

    Tratamiento de valores nulos o inconsistentes.

    Codificación de variables categóricas (One-Hot, Label Encoding).

    Normalización o estandarización de variables numéricas.

    Verificación de balance de clases (¿hay desbalance entre clientes que cancelan y los que no?).

3. Análisis de Correlación y Selección de Variables

    Matriz de correlación para variables numéricas.

    Análisis de importancia con métodos como SelectKBest, RFE, o árboles de decisión.

    Eliminación de variables irrelevantes o redundantes.

4. Entrenamiento de Modelos

    Separación en conjunto de entrenamiento y prueba (train_test_split).

    Entrenar al menos dos modelos de clasificación (por ejemplo: Random Forest, XGBoost, Logistic Regression).

    Ajuste de hiperparámetros si es necesario (GridSearchCV o RandomizedSearchCV).

5. Evaluación de Modelos

    Métricas clave: Accuracy, Precision, Recall, F1-Score, ROC-AUC.

    Matriz de confusión para visualizar errores.

    Comparación entre modelos para elegir el más robusto.

6. Interpretación y Conclusiones Estratégicas

    Análisis de importancia de variables (feature importance).

    Identificación de los factores más influyentes en la cancelación.

    Recomendaciones estratégicas para la empresa basadas en los hallazgos.